src/pcp/geolocate/pcp-geolocate.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2022-2023 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PCP-GEOLOCATE 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-geolocate\f1 \- discover collector system geographical labels
.SH SYNOPSIS
\f3pcp\f1 \f3geolocate\f1
.SH DESCRIPTION
.B pcp-geolocate
reports the latitude and longitude for the local Performance Co-Pilot
collector host in JSON format.
This geolocation information is sourced from the cache file
.I $PCP_SYSCONF_DIR/labels/optional/geolocate
if it exists, else an attempt is made to perform geolocation based
on the host IP address, via several online sources (REST APIs).
.PP
The output from this command is suited for storing as metric labels by
saving it to the cache file mentioned above.
.PP
The opt-in
.BR systemd (1)
service unit file for this command provides an automated location
discovery for PCP metric labels.
.SH METRIC LABELS
The JSON output is the preferred format for storing geographical location
as PCP metric labels.
Every PCP metric available from this host will be tagged with the labels
for latitude and longitude, enabling tools that make use of this such as
the Grafana geomap panel.
.PP
For more information about metric labels refer to the
.BR pmcd (1),
.BR pminfo (1),
.BR pmlogger (1)
and
.BR pmLookupLabels (3)
manual entries.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1),
.BR pminfo (1),
.BR pmseries (1),
.BR pmlogger (1),
.BR systemd (1),
.BR pmLookupLabels (3),
.BR pcp.conf (5)
and
.BR pcp.env (5).

.\" control lines for scripts/man-spell
.\" +ok+ Grafana geomap


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pcp/geolocate/pcp-geolocate.py
#!/usr/bin/pmpython
#
# Copyright (C) 2022-2023 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=broad-except, bare-except, protected-access
""" Display geographical location from local cache or IP lookup """

from pcp.pmapi import pmContext as PCP
try:
    import urllib.request as httprequest
except Exception:
    import urllib2 as httprequest
import threading
import time
import json
import sys
import os

start = time.time()
output = '{"latitude":%s,"longitude":%s}'
writing = False
stdoutfd = sys.stdout

# first extract from cached (labels) file
try:
    if len(sys.argv) == 2:  # use single argument if presented
        path = sys.argv[1]
    else:
        path = PCP.pmGetConfig('PCP_SYSCONF_DIR')
        path += '/labels/optional/geolocate'
    cached = json.load(path)
    print(output % (cached['latitude'], cached['longitude']))
except:
    try:
        sys.stdout = open(path, 'w')    # create (or overwrite)
        writing = True
    except PermissionError:
        sys.stderr.write('No permission to write to %s\n' % path)
        sys.stdout = stdoutfd


# setup threads for handling parallel REST API requests
threads = []

def finish(url):
    if writing:
        sys.stdout.close()
        sys.stdout = stdoutfd
    #print("'%s\' fetched in %ss" % (url, time.time() - start))
    os._exit(0) # not sys.exit, which awaits all threads


def ipinfo(url):
    with httprequest.urlopen(url) as http:
        js = http.read().decode('utf-8')
        data = json.loads(js)
        coords = data['loc'].split(',')
        print(output % (coords[0], coords[1]))
    finish(url)

IPINFO = "https://ipinfo.io/json"
threads.append(threading.Thread(target = ipinfo, args = (IPINFO, )))


def mozilla(url):
    with httprequest.urlopen(url) as http:
        js = http.read().decode('utf-8')
        data = json.loads(js)
        coords = data['location']
        print(output % (coords['lat'], coords['lng']))
    finish(url)

MOZILLA = "https://location.services.mozilla.com/v1/geolocate?key=geoclue"
threads.append(threading.Thread(target = mozilla, args = (MOZILLA,)))


try:
    for thread in threads:
        thread.start()
        thread.join()
except KeyboardInterrupt:
    finish('(none)')
